windows/src/main/csharp/ch/cyberduck/ui/core/preferences/PackagedDataMigrator.cs

```diff
@@ -0,0 +1,94 @@
+﻿using ch.cyberduck.core.preferences;
+using Ch.Cyberduck.Core.Preferences;
+using org.apache.logging.log4j;
+using System;
+using System.IO;
+using Windows.Win32.UI.Shell;
+using static Windows.Win32.CorePInvoke;
+using static Windows.Win32.PInvoke;
+
+namespace Ch.Cyberduck.Core.Ui.Preferences;
+
+internal static class PackagedDataMigrator
+{
+    public static void Migrate()
+    {
+        // Logger is used for short time.
+        // Don't need to keep an instance alive longer than necessary
+        var logger = LogManager.getLogger(typeof(PackagedDataMigrator).FullName);
+        var preferences = PreferencesFactory.get();
+        if (preferences.getBoolean("application.datadir.migrate"))
+        {
+            if (logger.isDebugEnabled())
+            {
+                logger.debug("Skipping DataDir migration");
+            }
+
+            return;
+        }
+
+        preferences.setProperty("application.datadir.migrate", true);
```

Comment:
Set only *after* migrating.

Reply:
"best-effort" for me is try once, if it fails for whatever reason, don't try again later.